### ISSUE TITLE
Memory leaks while analyzing

### ISSUE THREAD

**Issue by ging-dev** [phpactor/phpactor#2381]
Continue of https://github.com/phpactor/phpactor/issues/2377#issue-1909936536

Test with command: `phpactor worse:analyse src/Phan/Language`

![image](https://github.com/phpactor/phpactor/assets/42226341/7560fed8-76fb-4480-bea1-0f30236af19e)
![image](https://github.com/phpactor/phpactor/assets/42226341/6fa2b260-b287-4d0c-aefc-04acb57e52c9)

**Comment by dantleech** [comment]
This is very slow (for me probably 15 minutes). But it's not a good indicator for a memory leak as it's expected that this command will cache the results as it goes through _all_ the files in the source tree.

The `worse:analyse` command is really only for debugging and development.

What is the real problem?

**Comment by ging-dev** [comment]
> This is very slow (for me probably 15 minutes). But it's not a good indicator for a memory leak as it's expected that this command will cache the results as it goes through _all_ the files in the source tree.
> 
> The `worse:analyse` command is really only for debugging and development.
> 
> What is the real problem?

![image](https://github.com/phpactor/phpactor/assets/42226341/944001b9-98dd-444b-b887-eb0b9fb61f64)

It stuck at `Resolving code actions`, there are some signs of a memory leak occurring:
![image](https://github.com/phpactor/phpactor/assets/42226341/645681e5-8597-4b42-a4c8-0d7c78ffd8b5)

**Comment by dantleech** [comment]
The problem is that it's extremely slow, this class is 4k lines long with lots of complex types. So I think it's better to reclassify this as a performance issue.

**Comment by dantleech** [comment]
hmm, running this on master now:

```
120 problems in 40.4439 seconds with 564,884,936b memory
```

**Comment by ging-dev** [comment]
> hmm, running this on master now:
> 
> ```
> $ phpactor worse:analyse src/Phan/Language
> 120 problems in 40.4439 seconds with 564,884,936b memory
> ```
> 
> although I'm not sure what may have changed 🤔 it's still slow, but "usable".

should probably limit the number of `textDocuments/codeActions`, `vscode` probably sends too many of these requests.

**Comment by dantleech** [comment]
yeah, it probably makes sense to have a "process(request) manager" that can keep track of the running requests and _cancel_ "duplicated" ongoing requests.

VS code doesn't send cancellation requests for code-actions (but it does send them for completion).
